File: ticket-backend/src/main/java/aics/domain/hall/dtos/HallDto.java
package aics.domain.hall.dtos;

import aics.domain.hall.entities.Hall;
import aics.domain.hall.entities.Seat;
import aics.infrastructure.core.LabelValue;
import lombok.Data;
import lombok.experimental.Accessors;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Data
@Accessors(chain = true)
public class HallDto implements Serializable {
    private Long hallId;
    private String name;
    private String description;
    private int seatsRows;
    private int seatsColumns;
    private List<SeatListItemDto> seats = new ArrayList<>();
    private LabelValue<Long> providerRef;

    public static HallDto fromHall(Hall hall, List<Seat> seats) {
        if (hall == null) {
            return null;
        }
        LabelValue<Long> providerRef = hall.getProvider() != null
            ? new LabelValue<>(hall.getProvider().getName(), hall.getProvider().getProviderId())
            : null;
        return new HallDto()
            .setHallId(hall.getHallId())
            .setName(hall.getName())
            .setDescription(hall.getDescription())
            .setSeatsRows(hall.getSeatsRows())
            .setSeatsColumns(hall.getSeatsColumns())
            .setSeats(seats.stream().map(SeatListItemDto::fromSeat).collect(Collectors.toList()))
            .setProviderRef(providerRef);
    }
}

File: ticket-frontend/src/modules/hall/hall-dto.ts
import { LabelValue } from '../core/label-value';
import { TypeUtils } from '../core/type-utils';
import { SeatListItemDto } from './seat-list-item-dto';

export class HallDto {
    hallId: number | null = null;
    name: string = '';
    seatsRows: number = 0;
    seatsColumns: number = 0;
    description: string = '';
    seats: SeatListItemDto[] = [];
    providerRef: LabelValue<number> | null = null;

    static fromObj(obj: any): HallDto | null {
        if (!obj) {
            return null;
        }
        const hallDto: HallDto = new HallDto();
        hallDto.hallId = obj.hallId;
        hallDto.name = obj.name;
        hallDto.seatsRows = obj.seatsRows;
        hallDto.seatsColumns = obj.seatsColumns;
        hallDto.description = obj.description;
        hallDto.seats = SeatListItemDto.listFromObjList(obj.seats);
        hallDto.providerRef = LabelValue.fromObj(obj.providerRef);
        return hallDto;
    }

    static listFromObjList(objs: any[]): HallDto[] {
        if (!objs) {
            return [];
        }
        return objs.map(HallDto.fromObj).filter(TypeUtils.isNonNullable);
    }
}